### PR TITLE
Get a shopcart API

### PR DESCRIPTION
**Pull request by momo100110**
1. implement Get A Shopcart API in routes.py
2. add tests in test_models.py and test_routes.py
3. update README
image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -55,6 +55,18 @@ tests/              - test cases package
 └── test_routes.py  - test suite for service routes
 ```
 
+## RESTful routes for orders and items
```
Comment (wl2484) [inline README.md:58]: Given PR #36, the one with API doc template has been merged to master, would you like to merge master to resolve and add corresponding changes?

File: service/routes.py [hunk 2]
```diff
@@ -89,3 +91,76 @@ def create_shopcart():
...
+        )
+    app.logger.info("Returning shopcart: %s", shopcart.id)
+    return make_response(jsonify(shopcart.serialize()), status.HTTP_200_OK)
+
+
+    
+
+
+
+# # RETRIEVE A ITEM FROM SHOPCART
```
Comment (wl2484) [inline service/routes.py:128]: could you help remove the commented code lines not in used below?

File: tests/test_models.py [hunk 3]
```diff
@@ -35,18 +39,39 @@ def setUp(self):
         db.drop_all()  # clean up the last tests
         db.create_all()  # make our sqlalchemy tables
 
-    def tearDown(self):
-        """ This runs after each test """
-        db.session.remove()
-        db.drop_all()
+    # def tearDown(self):
```
Comment (wl2484) [inline tests/test_models.py:42]: should we keep this method?

File: tests/test_routes.py [hunk 4]
```diff
@@ -34,7 +34,10 @@ def tearDownClass(cls):
         db.session.close()
 
     def setUp(self):
-        """ This runs before each test """
+        """Runs before each test"""
+        db.session.query(Shopcart).delete()  # clean up the last tests
```
Comment (wl2484) [inline tests/test_routes.py:38]: line 38 and line 39 duplicate with line 42 and line 43, can you help remove them?

File: tests/test_routes.py [hunk 5]
```diff
@@ -91,3 +94,117 @@ def test_index(self):
...
+        self.assertEqual(resp.status_code, status.HTTP_200_OK)
+        data = resp.get_json()
+        self.assertEqual(data["id"], test_shopcart.id)
+
+    def test_get_shopcart_not_found(self):
+        """It should not Read a Shopcart that is not found"""
+        response = self.client.get(f"{BASE_URL}/0")
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+    
+    # @patch.object(Shopcart, 'get_by_id', MagicMock(side_effect=Exception("DBAPIErr")))
```
Comment (wl2484) [inline tests/test_routes.py:114]: could you help remove the commented code lines not in used below?

File: tests/test_routes.py [hunk 6]
```diff
@@ -91,3 +94,117 @@ def test_index(self):
...
+            json={
+                "items": []
+            },
+            content_type="application/json",
+        )
+        self.assertEqual(resp.status_code, status.HTTP_400_BAD_REQUEST)
+
+       
+
+    # def test_get_item(self):
```
Comment (wl2484) [inline tests/test_routes.py:178]: could you help remove the commented code lines not in used below?

File: service/routes.py [hunk 7]
```diff
@@ -89,3 +91,76 @@ def create_shopcart():
...
+@app.route("/shopcarts/<int:shopcart_id>", methods=["GET"])
+def get_shopcarts(shopcart_id):
+    """
+    Retrieve a single Shopcart
+    This endpoint will return an Shopcart based on its id
+    """
+    app.logger.info("Request for Shopcart with id: %s", shopcart_id)
+
+    # See if the shopcart exists and abort if it doesn't
+    # try:
```
Comment (wl2484) [inline service/routes.py:109]: I would vote for uncommenting line 109-112 😛

Comment (wl2484) [reply service/routes.py:109]: If you're worrying about the test case, please refer to this commit: https://github.com/CSCI-GA-2820-SU23-001/shopcarts/pull/37/commits/f03863a54c835ecacc44cc49a7e544ba54e0890a

The two would be almost the same because they're mocking the same issue.

**Review by wl2484 — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (wl2484) [reply README.md:58]: Also, there's conflicts with the master branch.

File: service/routes.py (hunk 7 above)
Comment (wl2484) [reply service/routes.py:109]: however, give this is not an requirement, it is fine if you leave it as is 🙂

File: README.md (hunk 1 above)
Comment (momo100110) [reply README.md:58]: Yes, sure! Let me fix the conflicts and commit again.

File: tests/test_models.py (hunk 3 above)
Comment (momo100110) [reply tests/test_models.py:42]: yes we can keep it

File: tests/test_routes.py (hunk 4 above)
Comment (momo100110) [reply tests/test_routes.py:38]: sure, sorry about the duplicate

File: service/routes.py (hunk 7 above)
Comment (momo100110) [reply service/routes.py:109]: There are some bugs when I use this try-catch structure when testing. I think maybe we can add if we have enough time after all the APIs are implemented.

Comment (wl2484) [reply service/routes.py:153]: OK

File: service/models.py [hunk 8]
```diff
@@ -134,11 +140,11 @@ def deserialize(self, data: dict) -> None:
                 f"Invalid {type(self).__name__}: failed to deserialize request body\nError message: {error}"
             ) from error
 
-    @classmethod
-    def get_by_id(cls, id):
-        """ Get shopcart by primary key: id """
-        logger.info(f"Get {cls.__name__} by id={id}")
-        return cls.query.get(id)
+    # @classmethod
```
Comment (wl2484) [inline service/models.py:143]: given get_by_id has been moved to the Base class, should we remove these lines?

File: service/models.py [hunk 9]
```diff
@@ -194,8 +200,8 @@ def deserialize(self, data: dict) -> None:
             data (dict): A dictionary containing the resource data
         """
         try:
-            # self.id = data["id"]
-            # self.shopcart_id = data["shopcart_id"]
+            self.id = data["id"]
```
Comment (wl2484) [inline service/models.py:203]: I think line 203 and 204 should be removed because Item.deserialize() would be called when adding an Item to shopcart, and at that item we wouldn't know the item id. shopcart id is not required to be present in the request body, either, because it is already carried in the API path (e.g. /shopcarts/{shopcart_id}/items)

File: tests/factories.py [hunk 10]
```diff
@@ -32,22 +33,24 @@ class Meta:
...
+
 class ItemFactory(factory.Factory):
     """Creates fake Items"""
 
     class Meta:
         model = Item
 
     id = factory.Sequence(lambda n: n)
-    shopcart_id = None
+    shopcart_id = factory.Sequence(lambda n: n)
```
Comment (wl2484) [inline tests/factories.py:53]: I think `shopcart_id` should not be randomly assigned. shopcart_id is always known when we're creating/updating/getting/deleting an item 🤔

File: tests/test_routes.py [hunk 11]
```diff
@@ -49,7 +51,8 @@ def tearDown(self):
 
     def _create_an_empty_shopcart(self, shopcart_count):
         """ Factory method to create empty shopcarts """
-        shopcarts = list()
+        #shopcarts = list()
+        shopcarts = []
```
Comment (wl2484) [inline tests/test_routes.py:55]: There's no difference between using `shopcarts = []` or `shopcarts = list()` here. They are identical lol

Comment (momo100110) [reply tests/test_routes.py:55]: Exactly lol

File: tests/factories.py (hunk 10 above)
Comment (momo100110) [reply tests/factories.py:53]: thank you for reminding this I'll modify this line and commit again

File: service/models.py (hunk 8 above)
Comment (momo100110) [reply service/models.py:143]: yes, i just commented these lines in my local branch. I'll remove them in Shopcart class.

File: service/routes.py (hunk 2 above)
Comment (momo100110) [reply service/routes.py]: Done

File: tests/test_routes.py (hunk 5 above)
Comment (momo100110) [reply tests/test_routes.py]: done

File: tests/test_routes.py (hunk 6 above)
Comment (momo100110) [reply tests/test_routes.py]: done

File: service/models.py (hunk 9 above)
Comment (momo100110) [reply service/models.py:203]: I have a question about why we wouldn't know the item id. Do you mean item id should be increased in the order they are added to a shopcart, i.e., one at a time, rather than being specified at the time of being added into one shopcart with some assumed id?

File: tests/test_routes.py [hunk 12]
```diff
@@ -92,6 +96,54 @@ def test_index(self):
...
+    def test_create_shopcarts(self):
+        """ It should create a shopcart """
+        shopcart = ShopcartFactory()
+        resp = self.client.post(
+            f"{BASE_URL}",json = shopcart.serialize(), content_type = "application/json"
+        )
+        
+        self.assertEqual(resp.status_code, status.HTTP_201_CREATED)
+        
+        location = resp.headers.get("location")
```
Comment (wl2484) [inline tests/test_routes.py:125]: given we do not set location in headers, line 125-126 can be removed 🙂

File: tests/test_routes.py [hunk 13]
```diff
@@ -92,6 +96,54 @@ def test_index(self):
...
+        data = resp.get_json()
+        self.assertEqual(data["id"], test_shopcart.id)
+
+    def test_get_shopcart_not_found(self):
+        """It should not Read a Shopcart that is not found"""
+        response = self.client.get(f"{BASE_URL}/0")
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+    
+   
+    def test_create_shopcarts(self):
```
Comment (wl2484) [inline tests/test_routes.py:116]: this seems to be a test for create_shopart, and according to Professor's comment in another channel, I feel that we should put it in another PR 🤔

File: tests/test_routes.py [hunk 14]
```diff
@@ -92,6 +96,54 @@ def test_index(self):
...
+        
+        location = resp.headers.get("location")
+        #self.assertIsNotNone(location,None)
+ 
+        new_shopcart = resp.get_json()
+        # self.assertEqual(data["date"], shopcart.date,"date does not match")
+        self.assertEqual(new_shopcart["name"],shopcart.name, "name does not match")
+        
+
+    def test_create_shopcart_missing_info(self):
```
Comment (wl2484) [inline tests/test_routes.py:133]: this seems to be a test for create_shopart, and according to Professor's comment in another channel, I feel that we should put it in another PR 🤔

File: service/models.py (hunk 8 above)
Comment (momo100110) [reply service/models.py:143]: Actually, I wrote these test cases in test_models.py just like this sample-accounts lol

File: tests/test_routes.py (hunk 13 above)
Comment (momo100110) [reply tests/test_routes.py:116]: ok

File: tests/test_routes.py (hunk 14 above)
Comment (momo100110) [reply tests/test_routes.py:133]: ok

File: service/models.py (hunk 9 above)
Comment (wl2484) [reply service/models.py:203]: Because when we call Create Shopcart Item API, we wouldn't know the new item's id until we actually insert it into DB

**Review by wl2484 — APPROVED** [review]
LGTM
